DS-3280 Fix HQL Syntax at HandleDAOImpl

FacundoAdorno opened the pull request:
Fix error that happens when executing `./dspace update-handle-prefix`. There were some errors in the HQL syntax at HandleDAOImpl class when updating the HANDLE table.

https://jira.duraspace.org/browse/DS-3280

FacundoAdorno committed:
DS-3280 Fix HQL Syntax at HandleDAOImpl

Solved the error when executing ./dspace update-handle-prefix
in commandline. There were some errors in the HQL syntax at
HandleDAOImpl class when updating the HANDLE table.

tdonohue commented:
+1 via code review. Looks like a rather obvious bug to me, but could use a quick test from someone else.

hardyoyo commented:
+1 by code review

abollini commented on dspace-api/src/main/java/org/dspace/handle/dao/impl/HandleDAOImpl.java:
```diff
@@ -83,7 +83,7 @@ public long countHandlesByPrefix(Context context, String prefix) throws SQLExcep
     public int updateHandlesWithNewPrefix(Context context, String newPrefix, String oldPrefix) throws SQLException
     {
-        String hql = "UPDATE handle set handle = concat(:newPrefix,'/',id WHERE handle like :oldPrefix%";
+        String hql = "UPDATE Handle set handle = concat(:newPrefix,'/',handle_id) WHERE handle like concat(:oldPrefix,'%')";
```
Not sure but the use of handle_id here looks strange to me.
I have also found this post on stackoverflow
http://stackoverflow.com/questions/27345133/use-column-name-instead-of-entity-property-in-jpql

tdonohue replied:
Aha, I think @abollini is correct.  This is HQL and not SQL. And, Hibernate does prefer using the name of the *entity* (in this case "id") instead of the name of the actual DB column (in this case "handle_id").  So, it probably is more correct to use "id" here.

FacundoAdorno replied:
@abollini @tdonohue you are right! I do not realize of this. I've replaced the "handle_id" by the "id" property and it works fine!

FacundoAdorno committed:
DS-3280 Replace 'handle_id' by 'id' in HQL query

The HQL query was fixed to use the 'id' of the Hibernate entity
instead of the 'handle_id' DB column.

abollini committed:
Merge branch 'DS-3280' of https://github.com/sedici/DSpace

abollini commented:
I'm testing the PR but I'm facing with two issues:
- update the metadatavalues throw an auhtorize exception (easy to fix putting a turnOffAuthorization)
- fixing the previous error I get
`Updating metadatavalues table... Exception: org.dspace.content.DSpaceObject_$$_jvst7c3_19 cannot be cast to org.dspace.content.Item
java.lang.ClassCastException: org.dspace.content.DSpaceObject_$$_jvst7c3_19 cannot be cast to org.dspace.content.Item
	at org.dspace.content.ItemServiceImpl.updateLastModified(ItemServiceImpl.java:1)
	at org.dspace.content.MetadataValueServiceImpl.update(MetadataValueServiceImpl.java:83)
	at org.dspace.handle.UpdateHandlePrefix.main(UpdateHandlePrefix.java:101)`

it looks like the retrieved Item is an hibernate proxy of DSpaceObject instead of Item so it cannot be cast to item.

abollini committed:
DS-3280 fix authorization and lazy proxy class cast issues

abollini commented:
I have fixed the two previous reported issues here

FacundoAdorno committed:
Merge branch '4Science-DS-3280' into DS-3280

FacundoAdorno commented:
@abollini I have tested it and works great! A few days ago I have detected the reported issues, and been trying to find a solution before report them. Thanks! Successfully merged!

With respect to the **UpdateHandlePrefix.java class**, I found another problem:

- When 'dc.identifier.uri' is updated at `metadatavalue` table , the "handle suffix" (/handle/{prefix}/**{suffix}**) is missing.

For example, when we run `./dspace update-handle-prefix 123456789 11746` and we have the dc.identifier.uri: "http://hdl.handle.net/123456789/2100", the update result is "http://hdl.handle.net/11746". 

The error is at [this line](https://github.com/DSpace/DSpace/blob/daa0af54d50adee97a798eec004c8e9fc3e7b222/dspace-api/src/main/java/org/dspace/handle/UpdateHandlePrefix.java#L99). In a moment I'll push the fix to this issue.

FacundoAdorno committed:
DS-3280 Fix inconsistency after running 'update-handle-prefix'

The 'handle suffix' in the text_value DB column is lost
after running the 'update-handle-prefix'.

lap82 commented:
+1 tested, it works

pnbecker commented:
merged after discussion in today's dev meeting (http://irclogs.duraspace.org/index.php?date=2016-09-21).
